### PR TITLE
fix: lexer positions

### PR DESCRIPTION
**Pull request by JonathanStarup**
`Lexer.retreat` didn't handle lines/columns properly and I couldn't comprehend how `State.end` was tracked.

Now i save the width of the previous line, making the end position computable on the spot when building tokens instead of keeping track of it dynamically. This width of the previous line needs to be recomputed if you retreat across multiple lines. In a follow up PR i think we should remove `retreat` and instead store checkpoints (line/col/offset/prevWidth) and just directly reset instead of going back one by one. So i judge this potential performance cost small and gone in the next PR.

These new positions agree with all locations of the existing lexer, except for some lines ending with an integer literal, in which the existing lexer gives a wrong location.

### CODE REVIEW

**Comment by JonathanStarup** [comment]
!!!jar

**Comment by JonathanStarup** [comment]
Also looks good with manual review

File: main/src/ca/uwaterloo/flix/language/phase/Lexer.scala [hunk 1]
```diff
@@ -175,15 +187,38 @@ object Lexer {
...
       s.current.column -= 1
-      s.end.column -= 1
     }
   }
 
+  /**
+    * Assuming that `s.current` is on the end of a line (i.e. on `\n`), recomputes the length of
+    * the previous line.
+    */
+  private def recomputePrevWidth()(implicit s: State): Unit = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lexer.scala:201]: Will this be slow?

**Comment by magnus-madsen** [comment]
How can we test this? It could break source locations in non-obvious ways.

**Comment by magnus-madsen** [comment]
I feel like we really need that source location checker...

**Comment by JonathanStarup** [comment]
I ran it with the old `s.end` location and checked that they agreed. Then manually checked a random subset of them. They were all on literal int tokens and related to end-of-line things on which the existing lexer has a bug. 

But yeah working forward we should have a way to test it

File: main/src/ca/uwaterloo/flix/language/phase/Lexer.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Lexer.scala:201]: It traverses a line, so somewhat. but 1) there is a (a little) less overhead during normal advance nad 2) I think we should delete retreat and do it with checkpoints, then this function will also be deleted 3) it is only triggered if you retreat across a line 4) it is stored in state, so its not necessarily spending effort. 5) i could do it on-demand if needed, but due to point 2 i don't think its worth it

**Comment by magnus-madsen** [comment]
I am not against, but I think I need a more in depth in person description.

**Review by mlutze — APPROVED** [review]
I approve on condition that we move to checkpoints afterward

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**